### ISSUE TITLE
DistributedSampler can't shuffle the dataset

### ISSUE THREAD

**Issue by elk-cloner** [huggingface/transformers#3721]
# 🐛 Bug

## Information

I'm trying to fine-tune BERT model using ```run_language_modeling.py```.  

Language I am using the model on is Persian:

The problem arises when using:
* [x] the official example scripts: (give details below)
* [ ] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: (give details below)

But according to this [issue](https://github.com/pytorch/pytorch/issues/31771) there is a bug in ```torch.utils.data.distributed.DistributedSampler``` so that during different epochs shuffling operation doesn't work properly(it's not working). 
To solve this problem: according to pytorch official example [here](https://github.com/pytorch/examples/blob/ad775ace1b9db09146cdd0724ce9195f7f863fff/imagenet/main.py#L238), we should add ```train_sampler.set_epoch(epoch)``` before each new epoch at this [line](https://github.com/huggingface/transformers/blob/f8208fa456039b46873a2e497b6318d30a4fc84e/examples/run_language_modeling.py#L322) 

## To reproduce

1. compare batches between different epoch like mentioned [issue](https://github.com/pytorch/pytorch/issues/31771) 

## Environment info

     
- `transformers` version: transformers==2.8.0
- Platform: Ubuntu 18.04
- Python version: 3.7
- PyTorch version (GPU?): torch==1.4.0 (Yes)
- Tensorflow version (GPU?): tensorflow-gpu==2.1.0 (Yes)
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: distributed

**Comment by julien-c** [comment]
I think you are right

**Comment by vlevit** [comment]
Isn't there the same issue in other places?
E.g. in trainer.py: https://github.com/huggingface/transformers/blob/97a375484c618496691982f62518130f294bb9a8/src/transformers/trainer.py#L305-L307

**Comment by julien-c** [comment]
I forgot to re-add this in Trainer when merging #3800 

It's on my todo-list, but feel free to open a PR if you can do it faster than I can

**Comment by vlevit** [comment]
Great. Personally I've not yet upgraded to the newer version with trainer.py, so I'll leave it for you, thanks.
